clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/loadbalancer/UpsertOpenstackLoadBalancerAtomicOperation.groovy

```diff
@@ -16,203 +16,390 @@
     task.updateStatus UPSERT_LOADBALANCER_PHASE, "Initializing upsert of load balancer ${description.id ?: description.name} in ${description.region}..."
-    String subnetId = description.subnetId
+
     String region = description.region
-    LoadBalancerPool newLoadBalancerPool = new LoadBalancerPool(
-      id: description.id,
-      name: description.name,
-      protocol: description.protocol,
-      method: description.method,
-      subnetId: subnetId,
-      internalPort: description.internalPort)
-    VirtualIP virtualIP = new VirtualIP(
-      name: description.name,
-      subnetId: subnetId,
-      poolId: description.id,
-      protocol: description.protocol,
-      port: description.externalPort)
-    PoolHealthMonitor poolHealthMonitor = description.healthMonitor
-
-    LbPool resultPool
+    LoadBalancerV2 resultLoadBalancer
+
     try {
-      if (newLoadBalancerPool.id) {
-        resultPool = updateLoadBalancer(region, newLoadBalancerPool, virtualIP, poolHealthMonitor)
+      if (!this.description.id) {
+        resultLoadBalancer = createLoadBalancer(region, description.name, description.subnetId)
       } else {
-        resultPool = createLoadBalancer(region, subnetId, newLoadBalancerPool, virtualIP, poolHealthMonitor)
+        resultLoadBalancer = clientProvider.getLoadBalancer(region, description.id)
       }
-    } catch (OpenstackProviderException ope) {
-      throw new OpenstackOperationException(AtomicOperations.UPSERT_LOAD_BALANCER, ope)
-    }
 
-    task.updateStatus UPSERT_LOADBALANCER_PHASE, "Done upserting load balancer ${resultPool?.name} in ${region}"
-    return [(region): [id: resultPool?.id]]
-  }
+      Map<String, ListenerV2> existingListenerMap = buildListenerMap(region, resultLoadBalancer)
+      Map<String, Listener> listenerMap = description.listeners.collectEntries([:]) { Listener current ->
+        [(getListenerKey(current.externalPort, current.externalProtocol.name(), current.internalPort, current.internalProtocol.name())): current]
+      }
+      Map<String, ListenerV2> listenersToUpdate = [:]
+      Map<String, Listener> listenersToAdd = [:]
+
+      listenerMap.entrySet()?.each { Map.Entry<String, Listener> entry ->
+        ListenerV2 foundListener = existingListenerMap.get(entry.key)
+        if (foundListener) {
+          listenersToUpdate.put(entry.key, foundListener)
+        } else {
+          listenersToAdd.put(entry.key, entry.value)
+        }
+      }
 
-  protected LbPool createLoadBalancer(String region, String subnetId, LoadBalancerPool newLoadBalancerPool, VirtualIP virtualIP, PoolHealthMonitor poolHealthMonitor) {
-    LbPool resultPool
-    OpenstackClientProvider openstackClientProvider = getClientProvider()
+      Set<String> listenersToDelete = Sets.difference(existingListenerMap.keySet(), Sets.union(listenersToAdd.keySet(), listenersToUpdate.keySet()))
 
-    newLoadBalancerPool.setCreatedTime(System.currentTimeMillis())
+      if (!listenersToDelete.isEmpty()) {
```

Comment:
As far as I know, `isEmpty` is only really used when you have a reason to disambiguate between `null` and `[:]` or `[]`. I don't think this is the case here, so it's more idiomatic to rely on groovy-truthiness, e.g. `if (!listenersToDelete) {`